source/openpulse/openpulse/parser.py

```diff
@@ -0,0 +1,303 @@
+        if ctx.WAVEFORM():
+            return openpulse_ast.WaveformType()
+        if ctx.PORT():
+            return openpulse_ast.PortType()
+        if ctx.FRAME():
+            return openpulse_ast.FrameType()
+
+    @span
+    def visitCalibrationBlock(self, ctx: openpulseParser.CalibrationBlockContext):
+        with self._push_context(ctx):
+            return openpulse_ast.CalibrationBlock(
+                body=[self.visit(statement) for statement in ctx.openpulseStatement()]
+            )
+
+    def visitScalarType(self, ctx: openpulseParser.ScalarTypeContext):
+        if ctx.WAVEFORM() or ctx.PORT() or ctx.FRAME():
+            return self._visitPulseType(ctx)
+        else:
+            return QASMNodeVisitor.visitScalarType(self, ctx)
+
+    def visitIndexOperator(self, ctx: openpulseParser.IndexOperatorContext):
+        if ctx.setExpression():
+            return self.visit(ctx.setExpression())
+
+        index_element = (
+            openpulseParser.ExpressionContext,
+            openpulseParser.RangeExpressionContext,
+        )
+
+        def predicate(child):
+            return isinstance(child, index_element)
+
+        return [self.visit(child) for child in ctx.getChildren(predicate=predicate)]
+
+    @span
+    def visitRangeExpression(self, ctx: openpulseParser.RangeExpressionContext):
+        # start, end and step are all optional as in [:]
+        # It could be [start:end] or [start:step:end]
+        start = None
+        end = None
+        step = None
+        colons_seen = 0
+
+        for child in ctx.getChildren():
+            if isinstance(child, openpulseParser.ExpressionContext):
+                expression = self.visit(child)
+                if colons_seen == 0:
+                    start = expression
+                elif colons_seen == 1:
+                    end = expression
+                else:
+                    step = end
+                    end = expression
+            elif child.getText() == ":":
+                colons_seen += 1
+
+        return ast.RangeDefinition(start=start, end=end, step=step)
+
+    @span
+    def visitReturnStatement(self, ctx: qasm3Parser.ReturnStatementContext):
+        if not (self._in_subroutine() or self._in_defcal):
+            _raise_from_context(ctx, "'return' statement outside subroutine or defcal")
+
+        if ctx.expression():
+            expression = self.visit(ctx.expression())
+        elif ctx.measureExpression():
+            expression = self.visit(ctx.measureExpression())
+        else:
+            expression = None
+        return ast.ReturnStatement(expression=expression)
+
+
+# Reuse some QASMNodeVisitor methods in OpenPulseNodeVisitor
+# The following methods are overridden in OpenPulseNodeVisitor and thus not imported:
+"""
+    "visitChildren",
+    "visitErrorNode",
+    "visitTerminal",
+    "visitArrayLiteral",
+    "visitCalibrationGrammarStatement",
+    "visitCalStatement",
+    "visitDefcalStatement",
```

Comment:
**issue**: I don't see any of these overridden in `OpenPulseNodeVisitor`.

Reply:
Indeed. Forgot to update the comments. Fixed.